include/swift/Immediate/SwiftMaterializationUnit.h

```diff
@@ -0,0 +1,176 @@
+//===--- SwiftMaterializationUnit.h - JIT Swift ASTs ------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Defines the `SwiftMaterializationUnit` class, which allows you to JIT
+// individual Swift AST declarations.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IMMEDIATE_SWIFTMATERIALIZATIONUNIT_H
+#define SWIFT_IMMEDIATE_SWIFTMATERIALIZATIONUNIT_H
+
+#include <memory>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ExecutionEngine/JITLink/JITLink.h"
+#include "llvm/ExecutionEngine/Orc/Core.h"
+#include "llvm/ExecutionEngine/Orc/EPCIndirectionUtils.h"
+#include "llvm/ExecutionEngine/Orc/IndirectionUtils.h"
+#include "llvm/ExecutionEngine/Orc/LLJIT.h"
+#include "llvm/ExecutionEngine/Orc/LazyReexports.h"
+#include "llvm/ExecutionEngine/Orc/SymbolStringPool.h"
+
+#include "swift/AST/TBDGenRequests.h"
+
+namespace swift {
+
+class CompilerInstance;
+
+/// A JIT stack able to lazily JIT Swift programs
+class SwiftJIT {
+public:
+  SwiftJIT(const SwiftJIT &) = delete;
+  SwiftJIT(SwiftJIT &&) = delete;
+  SwiftJIT &operator=(const SwiftJIT &) = delete;
+  SwiftJIT &operator=(SwiftJIT &&) = delete;
+
+  /// Attempt to create and initialize a new `SwiftJIT` with lazy compilation
+  /// enabled and an attached generator to search for symbols defined in the
+  /// current process.
+  static llvm::Expected<std::unique_ptr<SwiftJIT>> Create(CompilerInstance &CI);
+
+  /// Adds a plugin that will rename function symbols for lazy reexports.
+  /// Should be called only once.
+  void addRenamer();
+
+  ~SwiftJIT();
+
+  /// Get the dylib associated with the main program
+  llvm::orc::JITDylib &getMainJITDylib();
+
+  /// Register a the materialization unit `MU` with the `JITDylib``JD` and
+  /// create lazy reexports for all functions defined in the interface of `MU`
+  llvm::Error addSwift(llvm::orc::JITDylib &JD,
+                       std::unique_ptr<llvm::orc::MaterializationUnit> MU);
```

Comment:
In the future we should change this to construct the `MaterializationUnit` internally, but I think this works for now.